File: archive_detect/backend/llm_service.py
"""
LLM 服务（极简版，仅供"文件留底检测"使用）。

只暴露：
  - load_config()        : 启动时调用，读 ../config.json
  - detect_archival(text, user_prompt) : 调 LLM 给出留底判定 JSON

完整版（含证件解析、模板填写等）在主项目 backend/llm_service.py，本子项目用不到。
"""

import os
import json
import re
import time
from openai import OpenAI


CONFIG_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "config.json")
CONFIG: dict = {}

ARCHIVE_DETECT_INPUT_LIMIT_CHARS = 30000


def load_config():
    """从子项目根目录的 config.json 加载配置。"""
    global CONFIG
    if not os.path.exists(CONFIG_PATH):
        raise FileNotFoundError(f"配置文件不存在: {CONFIG_PATH}（请复制 config.example.json 为 config.json 并填入 LLM key）")
    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        CONFIG = json.load(f)


def _call_llm(prompt: str, max_retries: int = 3) -> str:
    """调用大模型 API，返回原始响应文本。429 退避 2/4/6s。"""
    llm = CONFIG.get("llm", {})
    api_key = llm.get("api_key", "")
    base_url = llm.get("base_url", "")
    model = llm.get("model", "")
    temperature = llm.get("temperature", 0.2)

    if not api_key or not base_url or not model:
        raise RuntimeError("LLM 配置不完整，请检查 config.json 的 llm.api_key / base_url / model")

    client = OpenAI(api_key=api_key, base_url=base_url)

    last_exc = None
    for attempt in range(max_retries):
        try:
            resp = client.chat.completions.create(
                model=model,
                messages=[{"role": "user", "content": prompt}],
                temperature=temperature,
            )
            raw = (resp.choices[0].message.content or "").strip()
            # 去掉 ``` 围栏与可能的 <tool_call> 残留
            raw = re.sub(r"^```(?:json)?\s*|\s*```$", "", raw, flags=re.IGNORECASE)
            raw = re.sub(r"<tool_call>.*?</tool_call>", "", raw, flags=re.DOTALL)
            return raw.strip()
        except Exception as e:
            last_exc = e
            msg = str(e).lower()
            if "429" in msg or "rate" in msg or "throttle" in msg:
                wait = 2 * (attempt + 1)
                print(f"[llm] 429/限流，{wait}s 后重试 ({attempt + 1}/{max_retries})...")
                time.sleep(wait)
                continue
            raise
    raise last_exc or RuntimeError("LLM 调用失败（达到最大重试次数）")


def _build_archive_detect_prompt(text: str, user_prompt: str) -> str:
    """把用户输入的多行判定标准拼接进 LLM prompt。"""
    doc_types = CONFIG.get("document_types") or [
        "身份证件", "学历证明", "婚姻证明", "财务证明", "工作证明",
        "申请表单", "合同协议", "报告说明", "简历", "其他",
    ]
    cat_str = " / ".join(doc_types)

    return (
        "你是一个文档归档（留底）判定助手。下面是用户描述的判定标准（请严格按它判定，未提到的维度不要发挥）：\n"
        "---用户判定标准开始---\n"
        f"{user_prompt}\n"
        "---用户判定标准结束---\n\n"
        "请阅读以下文件内容，判定该文件是否符合上述判定标准，并输出：\n"
        "1. is_archival: true / false\n"
        "2. confidence: 0-100 整数（基于文本证据强度）\n"
        "3. reason: 30-120 字判断依据，仅引用与判定标准直接相关的内容；"
        "**不要泄露金额、电话、身份证号、银行卡号、账号等敏感信息，遇到这类内容请用 [金额]/[手机号]/[身份证]/[银行卡] 等占位词代替**\n"
        "4. key_points: 3-6 条要点 bullets，遵循同样的脱敏要求\n"
        f"5. doc_category: 从下列中选一个：{cat_str}\n\n"
        "返回严格 JSON，不要 markdown 代码块、不要任何额外解释：\n"
        '{"is_archival": true, "confidence": 0, "reason": "...", "key_points": ["..."], "doc_category": "..."}\n\n'
        "文件内容：\n---\n"
        f"{text}\n---\n"
    )


def detect_archival(text: str, user_prompt: str) -> dict:
    """以用户提供的判定标准判断 text 是否符合留底要求。

    入参：
      text        - 已抽取的纯文本（OCR 或 docx 抽取）
      user_prompt - 用户输入的多行判定标准（必填）

    返回：
      {is_archival: bool, confidence: int 0-100, reason: str,
       key_points: list[str], doc_category: str}

    异常：ValueError - 文本为空 / user_prompt 为空 / LLM 返回非 JSON
    """
    if not text or not text.strip():
        raise ValueError("文件内容为空，无法判定")
    if not user_prompt or not user_prompt.strip():
        raise ValueError("判定标准 user_prompt 不能为空")

    src = text.strip()
    if len(src) > ARCHIVE_DETECT_INPUT_LIMIT_CHARS:
        head_n = ARCHIVE_DETECT_INPUT_LIMIT_CHARS // 2
        tail_n = ARCHIVE_DETECT_INPUT_LIMIT_CHARS - head_n
        src = (
            src[:head_n]
            + f"\n\n...[省略 {len(text) - ARCHIVE_DETECT_INPUT_LIMIT_CHARS} 字]...\n\n"
            + src[-tail_n:]
        )

    prompt = _build_archive_detect_prompt(src, user_prompt.strip())
    raw = _call_llm(prompt)

    # 解析 JSON（容错）
    try:
        start = raw.find("{")
        end = raw.rfind("}")
        if start >= 0 and end > start:
            data = json.loads(raw[start:end + 1])
        else:
            data = json.loads(raw)
    except json.JSONDecodeError as e:
        raise ValueError(f"LLM 返回非合法 JSON：{raw[:200]}") from e

    is_archival = data.get("is_archival")
    if isinstance(is_archival, str):
        is_archival = is_archival.strip().lower() in ("true", "1", "yes", "是")
    is_archival = bool(is_archival)

    try:
        confidence = int(data.get("confidence", 0))
    except (TypeError, ValueError):
        confidence = 0
    confidence = max(0, min(100, confidence))

    return {
        "is_archival": is_archival,
        "confidence": confidence,
        "reason": str(data.get("reason", "")).strip(),
        "key_points": [str(x).strip() for x in (data.get("key_points") or []) if str(x).strip()],
        "doc_category": str(data.get("doc_category", "其他")).strip() or "其他",
    }

File: frontend2/DocReview.ArchiveDetect/Services/LlmService.cs
using System.Text;
using System.Text.Json;

namespace DocReview.ArchiveDetect.Services;

// 复刻 Python backend/llm_service.py 的 detect_archival + _build_archive_detect_prompt。
// prompt 文案逐字迁移,分类体系 ARCHIVE_CATEGORIES_FULL 原样保留(业务知识,与语言无关)。
public class LlmService
{
    private readonly HttpClient _http;
    private readonly string _apiKey;
    private readonly string _baseUrl;
    private readonly string _model;
    private readonly double _temperature;
    private readonly ILogger<LlmService> _log;

    private const int InputLimitChars = 12000;

    public LlmService(HttpClient http, IConfiguration cfg, ILogger<LlmService> log)
    {
        _http = http;
        _log = log;
        _apiKey = cfg["Llm:ApiKey"] ?? "";
        _baseUrl = (cfg["Llm:BaseUrl"] ?? "").TrimEnd('/');
        _model = cfg["Llm:Model"] ?? "";
        _temperature = double.TryParse(cfg["Llm:Temperature"], out var t) ? t : 0.1;
    }

    // 公司售后留底分类体系(硬编码,逐字来自 llm_service.ARCHIVE_CATEGORIES_FULL)
    private const string ArchiveCategoriesFull = """
【递交前阶段应上传的 5 大类】
A. 客户基础文件:护照、身份证、中文信息表、个人简历、出生证明类文件、户口本、毕业证书&学位证书、结婚证&离婚证等婚姻状态文件、港澳通行证、房产证、工作证明信等
B. 客户个人文件:职业&专业证书、照片、社保/个税、不反对移民申明、无犯罪记录证明、体检类文件、地址证明、在读证明、资产/资信文件、公证/认证文件、成就类文件、学生签证、成绩单、疫苗本、翻译类文件、证书或获奖记录、客户访校指南、单身证明、个人文件等
C. 客户公司文件:营业执照、章程、股东会决议、验资报告、公司财报/审计报告、银行流水、业务合同/合作协议、组织架构图、股东名册、雇佣合同、公司介绍、办公室照片、办公室租赁合同等
D. 其他备用文件:律师文件、投资文件、批复留底文件、使馆申请表、合规部KYC留底文件、开户KYC文件、投资证明、购房文件、入境处申请表格、劳工卡申请、正签信、工作签证留底、商业计划书等
E. 转款凭证:服务过程中涉及的转账凭证等

【递交后阶段应上传的 4 大类】
F. 文案制作的递交文件类:递交全套留底、递交后补料留底等
G. 获批/失败:筛选/名额通知函、使馆信、入境处信件、批复函、录取通知、补料信、打款通知、体检通知、获身份文件、拒签信、撤案信等任何客户相关批复类文件
H. 其他文件:客户获批后协助客户留存的重要证件信息,如更新护照、入境小白条等
I. 停滞/放弃类文件:客户明确表示撤案、不再继续办理、不启动了/放弃办理了等主观原因不再继续办理项目的邮件、聊天等截图
""";

    public record ArchivalResult(string Verdict, int MatchScore, bool IsArchival,
        int Confidence, string Reason, List<string> KeyPoints, string DocCategory);

    public async Task<ArchivalResult> DetectArchivalAsync(
        string text, string userPrompt, string? stage, string? clientName, string? handler,
        CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(text)) throw new ArgumentException("文件内容为空,无法判定");
        if (string.IsNullOrWhiteSpace(userPrompt)) throw new ArgumentException("判定标准不能为空");

        var src = text.Trim();
        if (src.Length > InputLimitChars)
        {
            int head = InputLimitChars / 2, tail = InputLimitChars - head;
            src = src[..head] + $"\n\n...[省略 {text.Length - InputLimitChars} 字]...\n\n" + src[^tail..];
        }

        var prompt = BuildPrompt(src, userPrompt.Trim(), stage, clientName, handler);
        var raw = await CallLlmAsync(prompt, ct);

        // 容错解析 JSON:取首个 { 到末个 }
        int s = raw.IndexOf('{'), e = raw.LastIndexOf('}');
        var jsonStr = (s >= 0 && e > s) ? raw[s..(e + 1)] : raw;
        JsonElement data;
        try { data = JsonDocument.Parse(jsonStr).RootElement; }
        catch (JsonException) { throw new InvalidOperationException($"LLM 返回非合法 JSON: {raw[..Math.Min(200, raw.Length)]}"); }

        var verdict = (data.TryGetProperty("verdict", out var v) ? v.GetString() ?? "" : "").Trim().ToLower();
        if (verdict is not ("match" or "partial" or "mismatch")) verdict = "mismatch";

        int score = 0;
        if (data.TryGetProperty("match_score", out var ms))
        {
            if (ms.ValueKind == JsonValueKind.Number) ms.TryGetInt32(out score);
            else int.TryParse(ms.GetString(), out score);
        }
        score = Math.Max(0, Math.Min(100, score));

        var reason = data.TryGetProperty("reason", out var r) ? (r.GetString() ?? "").Trim() : "";
        var docCat = data.TryGetProperty("doc_category", out var dc) ? (dc.GetString() ?? "").Trim() : "";
        if (string.IsNullOrEmpty(docCat)) docCat = "其他";

        var keyPoints = new List<string>();
        if (data.TryGetProperty("key_points", out var kp) && kp.ValueKind == JsonValueKind.Array)
            foreach (var item in kp.EnumerateArray())
            {
                var val = item.GetString()?.Trim();
                if (!string.IsNullOrEmpty(val)) keyPoints.Add(val);
            }

        return new ArchivalResult(verdict, score, verdict == "match", score, reason, keyPoints, docCat);
    }

    private static string BuildPrompt(string text, string userPrompt, string? stage, string? clientName, string? handler)
    {
        var stageLabel = stage == "pre_submit" ? "递交前" : "递交后";
        var stageHint = (stage == "pre_submit" || stage == "post_submit")
            ? $"\n当前阶段: {stage} ({stageLabel})\n" : "";

        var nameParts = new List<string>();
        if (!string.IsNullOrEmpty(clientName)) nameParts.Add($"客户姓名：{clientName}");
        if (!string.IsNullOrEmpty(handler)) nameParts.Add($"办理人：{handler}");
        var nameHeader = nameParts.Count > 0 ? "本进展" + string.Join("；", nameParts) + "。\n" : "";

        var handlerClause = !string.IsNullOrEmpty(handler)
            ? "官方回执、递交确认、受理通知等文件上的「申请人」姓名通常对应本进展办理人而非客户本人,若文件上姓名与客户不一致但与办理人(含拼音/英文转写)一致,视为本进展有效留底,不判 mismatch。"
            : "";

        return
            "你是一个公司文件留底审核助手。请根据下方公司分类标准 + 用户判定提示词,审核文件。\n\n" +
            "---用户判定标准开始---\n" + userPrompt + "\n---用户判定标准结束---\n\n" +
            "---公司分类标准开始---\n" + stageHint + ArchiveCategoriesFull + "\n---公司分类标准结束---\n\n" +
            "重要判定指南:\n" + nameHeader +
            "- 本次审核只判断文件是否与公司留底分类体系相关,不核对具体项目名称、投资金额、转账金额等细节。同一客户的文件集合中,属于该客户配偶、子女、父母、共同申请人的文件也视为相关。不要因为文件上的人名与客户姓名不一致而判为 mismatch。只要文件内容可归入分类体系且与客户/办理人相关,即视为符合。\n" +
            "- **文件中出现的姓名若匹配客户姓名、办理人姓名或用户判定标准里列出的其他关联人(家庭成员/共同申请人等),包括其拼音或英文转写,即视为与本进展强关联,应判 match 或 partial,不要因当事人姓名与客户本人不一致就判 mismatch。**授权委托书(Power of Attorney)、公证认证文件等,其委托人/当事人常为客户的家属或本进展的办理人,属有效留底;" + handlerClause + "\n" +
            "- 如果一份文件明显是某类证件的标准格式(如身份证、护照、房产证),即使 OCR 提取质量差、部分文字乱码,也应判为 match 或 partial,不要因 OCR 噪声判 mismatch。\n" +
            "- **服务合同、客户确认类聊天记录/邮件/截图属于有效留底证据**,它们能证明相关服务已启动或客户已确认事项。合同归入 C 类,客户确认/放弃类沟通归入对应分类,**不要因为它们不是标准证件格式就判 mismatch**。\n" +
            "- 若文件体现了服务已启动或客户已确认某项服务内容,请在 key_points 中明确标注,供总体判定参考。\n" +
            "- **阶段错配不作硬性否决**:某文件更适用于其他阶段,只要文件本身可归入分类体系且与客户/办理人相关,最多判 partial,不要仅因阶段不符就判 mismatch。\n\n" +
            "请输出:\n" +
            "1. verdict: 三选一 (\"match\"|\"partial\"|\"mismatch\")\n" +
            "2. match_score: 0-100 整数\n" +
            "3. doc_category: 字母编号+子类名, 如 \"A-护照\"、\"G-批复函\"\n" +
            "4. reason: 30-120 字判断依据,遇敏感信息用 [金额]/[手机号]/[身份证]/[银行卡] 占位\n" +
            "5. key_points: 3-6 条要点\n\n" +
            "返回严格 JSON,不要 markdown 代码块:\n" +
            "{\"verdict\": \"match\", \"match_score\": 0, \"doc_category\": \"A-护照\", \"reason\": \"...\", \"key_points\": [\"...\"]}\n\n" +
            "文件内容：\n---\n" + text + "\n---\n";
    }

    private async Task<string> CallLlmAsync(string prompt, CancellationToken ct)
    {
        var body = new
        {
            model = _model,
            temperature = _temperature,
            messages = new[] { new { role = "user", content = prompt } }
        };
        using var req = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/chat/completions");
        req.Headers.Add("Authorization", $"Bearer {_apiKey}");
        req.Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        using var resp = await _http.SendAsync(req, ct);
        var respText = await resp.Content.ReadAsStringAsync(ct);
        if (!resp.IsSuccessStatusCode)
            throw new InvalidOperationException($"LLM HTTP {(int)resp.StatusCode}: {respText[..Math.Min(300, respText.Length)]}");

        var root = JsonDocument.Parse(respText).RootElement;
        return root.GetProperty("choices")[0].GetProperty("message").GetProperty("content").GetString() ?? "";
    }
}
